libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/Trace.hs

```diff
@@ -0,0 +1,283 @@
+-- We show the type family instances here for reference.
+{-
+instance STS (MOCKCHAIN era)
+  where
+  type State (MOCKCHAIN era) = MockChainState era
+  type Signal (MOCKCHAIN era) = (MockBlock era,UTxO era)
+  type Environment (MOCKCHAIN era) = ()
+-}
+-- ==============================================================
+
+newtype Gen1 era = Gen1 (Vector [Core.Tx era])
+
+instance
+  ( STS (MOCKCHAIN era),
+    Reflect era
+  ) =>
+  HasTrace (MOCKCHAIN era) (Gen1 era)
+  where
+  type BaseEnv (MOCKCHAIN era) = Globals
+
+  interpretSTS globals act = runIdentity $ runReaderT act globals
+
+  envGen _gstate = pure ()
+
+  sigGen (Gen1 txss) () mcs@(MockChainState newepoch (SlotNo lastSlot) count) = do
+    let NewEpochState _epoch _ _ _ _ pooldistr _ = newepoch
+    issuerkey <- chooseIssuer pooldistr
+    nextSlotNo <- (SlotNo . (+ lastSlot)) <$> choose (15, 25)
+    let txs = Seq.fromList (txss ! count)
+    -- Assmble it into a MockBlock
+    let mockblock = MockBlock issuerkey nextSlotNo txs
+    -- Run the STS Rules for MOCKCHAIN with generated signal
+    goSTS
+      (MOCKCHAIN reify)
+      ()
+      mcs
+      (mockblock)
+      ( \case
+          Left pdfs -> error ("MOCKCHAIN sigGen:\n" <> show (ppList (ppMockChainFailure reify) pdfs))
+          Right _mcs -> pure (mockblock)
+      )
+
+  shrinkSignal _x = []
+
+mapProportion :: (v -> Int) -> Map.Map k v -> Gen k
+mapProportion toInt m = frequency [(toInt v, pure k) | (k, v) <- Map.toList m]
+
+chooseIssuer :: PoolDistr crypto -> Gen (KeyHash 'StakePool crypto)
+chooseIssuer (PoolDistr m) = mapProportion getInt m
+  where
+    getInt x = floor (individualPoolStake x * 1000)
+
+-- ===================================================================================
+
+-- =========================================================================
+-- Making and running traces
+-- let gensize =  (def {oldUtxoPercent = 15})
+
+makeMOCKCHAINtrace ::
+  forall era.
+  (Reflect era, HasTrace (MOCKCHAIN era) (Gen1 era)) =>
+  Proof era ->
+  GenSize ->
+  Word64 ->
+  Gen (Trace (MOCKCHAIN era))
+makeMOCKCHAINtrace proof gensize tracelen = do
+  (vs, genstate) <- genTxSeq proof gensize tracelen
+  traceFromInitState @(MOCKCHAIN era)
+    testGlobals
+    (fromIntegral (length vs))
+    (Gen1 (Vector.map (map snd) vs))
+    (Just (\_ -> Right <$> genMockChainState proof genstate))
+
+chainTest ::
+  forall era.
+  ( Reflect era,
+    HasTrace (MOCKCHAIN era) (Gen1 era)
+  ) =>
+  Proof era ->
+  Word64 ->
+  GenSize ->
+  TestTree
+chainTest proof n gsize = testCaseSteps message action
+  where
+    message = "MockChainTrace in the (" ++ show proof ++ ") era."
+    action step = do
+      step ("Generating a sequence of Tx of length " ++ show n)
+      (vs, genstate) <- generate $ genTxSeq proof gsize n
+      step ("Applying the STS rules to the trace with " ++ show (length vs) ++ " blocks.")
+      !_trace <-
+        generate $
```

Comment:
This will not work. We cannot use `generate` in tests!!! It is not reproducable.

If you'll run into trouble figuring out how to avoid `generate` let me know, I'll try to dive into this code in a little bit more detail and will take care of it.

Reply:
Lets talk about this